File: BFS & DFS/79 Word Search/solution2.py
class Solution:
    def exist(self, board, word):
        """
        :type board: List[List[str]]
        :type word: str
        :rtype: bool
        """
        self.found = False
        for i in range(len(board)):
            for j in range(len(board[0])):
                self.visited = []
                self.dfs(i, j, board, word)
                if self.found == True:
                    return True
        return False

    def dfs(self, i, j, board, word):
        if len(word) == 0:
            self.found = True
            return

        if i < 0 or i >= len(board) or j < 0 or j >= len(board[0]) or word[0] != board[i][j] or (
                (i, j) in self.visited):
            return

        self.visited.append((i, j))

        self.dfs(i - 1, j, board, word[1:])
        self.dfs(i + 1, j, board, word[1:])
        self.dfs(i, j - 1, board, word[1:])
        self.dfs(i, j + 1, board, word[1:])

        if not self.found:
            self.visited.pop()

if __name__ == '__main__':
    board = [['A', 'B', 'C', 'E'],
             ['S', 'F', 'C', 'S'],
             ['A', 'D', 'E', 'E']]

    word = "ABCCED"
    assert(Solution().exist(board, word) == True)

    word = "SEE"
    assert(Solution().exist(board, word) == True)

    word = "ABCB"
    assert(Solution().exist(board, word) == False)

File: 36 Valid Sudoku/solution2.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        int m,n,t,l;
        for (int i=0;i<9;++i) {
            for (int j=0;j<8;++j) {
                for (int k=j+1;k<9;++k) {
                    m=i/3*3+j/3; n=i%3*3+j%3; t=i/3*3+k/3; l=i%3*3+k%3;
                    if ((board[i][k]!='.' && board[i][k] == board[i][j]) || 
						(board[k][i]!='.' && board[k][i] == board[j][i]) || 
						(board[t][l]!='.' && board[m][n] == board[t][l])) 
                        return false;
                }
            }
        }
        return true;
    }
};